### PR TITLE
Update to latest rules_haskell for Bazel 0.21 support

### PR DESCRIPTION
**Pull request by mjrussell**
I tried out hazel for the first time today, trying to get it running on an example project. It looks really promising!

Because I installed Bazel for the first time, it defaulted to 0.21 which required me to point to master of rules_haskell (https://github.com/tweag/rules_haskell/commit/4c81b7a8921db75828cbf002832d1b403c774eef).

Looks like Hazel doesn't support some of the breaking changes since the last release version so I spent some time learning the codebase and fixing them. Hope it's helpful!

### CODE REVIEW

File: BUILD.ghc [hunk 1]
```diff
@@ -38,10 +39,14 @@ cc_library(
...
+c2hs_toolchain(
+  name ="c2hs",
+  c2hs = "@c2hs//:bin",
+)
+
 haskell_toolchain(
     name = "ghc",
-    c2hs = "@c2hs//:bin",
-    version = "8.2.2",
+    version = "8.6.3",
```
Comment (mjrussell) [inline BUILD.ghc]: I was working ghc 8.6.3 on my own project...i can revert this change if you don't want it. Im a little dissapointed with the hardcoding of the ghc version in both the BUILD and WORKSPACE, but it doesn't look like there are great options.

File: BUILD.ghc [hunk 2]
```diff
@@ -2,6 +2,7 @@
 package(default_visibility = [ "//visibility:public" ])
 
 load("@io_tweag_rules_haskell//haskell:haskell.bzl", "haskell_toolchain")
+load("@io_tweag_rules_haskell//haskell:c2hs.bzl", "c2hs_toolchain")
```
Comment (mjrussell) [inline BUILD.ghc]: The majority of the changes were related to moving the c2hs_toolchain into its own import

File: hazel.bzl [hunk 3]
```diff
@@ -193,7 +193,7 @@ def hazel_custom_package_hackage(
   )
   http_archive(
     name = hazel_workspace(package_name),
-    build_file = "//third_party/haskell:BUILD.{0}".format(package_name),
+    build_file = "@ai_formation_hazel//third_party/haskell:BUILD.{0}".format(package_name),
```
Comment (mjrussell) [inline hazel.bzl]: I had an issue here as well, using `//` seemed to default to the project I was using Hazel in, not Hazel itself. This seemed to work for me though. I think this was mentioned as an issue in https://github.com/FormationAI/hazel/pull/54

Comment (thumphries) [reply hazel.bzl]: This is intentional - Hazel is not intended to be the repository for custom BUILD files, your repository is.

Comment (thumphries) [reply hazel.bzl]: We have discussed this before [here](https://github.com/FormationAI/hazel/issues/48) - maybe a documentation fix is in order

File: BUILD.ghc (hunk 1 above)
Comment (thumphries) [reply BUILD.ghc]: There's a bit of weirdness around the compilation of Hazel-internal cabal2bazel binary - see https://github.com/FormationAI/hazel/issues/53 for related discussion

Comment (mjrussell) [reply BUILD.ghc]: So I guess in a similar vein to https://github.com/FormationAI/hazel/pull/65#discussion_r249980599, should I be using a BUILD.ghc defined in my own repo? It wasn't totally clear me to how much of the WORKSPACE/BUILD items from the hazel repo I needed to get started. Without this bindist I couldn't get anything packages not bundles with GHC to compile

Comment (thumphries) [reply BUILD.ghc]: Yeah correct - I am not too familiar with this part of hazel but AFAIK a BUILD.ghc file will be necessary to set up the toolchain

Thanks for puzzling through all this and surfacing feedback. We really need to stand up a rules_haskell / hazel quick start repository keyed to a recent stackage snapshot.

Comment (mjrussell) [reply BUILD.ghc]: Ok, maybe I can take a shot at a basic repo with hazel. Im still unable to compile any c2hs code properly...something about my setup isn't picking up my mac's SDK root (/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/) although its finding the threaded-rts and unix-includes properly. Im not sure how to solve this so maybe throwing up a sample repo and you all can help me figure out where its going wrong.

The one in the rules_haskell_examples doesn't seem to be kept up to date and doesn't include any packages from stackage I think

Comment (mjrussell) [reply BUILD.ghc]: nvm was the annoying `BAZEL_USE_CPP_ONLY_TOOLCHAIN=1`

**Comment by mjrussell** [comment]
@thumphries I removed the parts that should be checked into the repo using hazel. Let me know if there's anything else you want done in this PR

**Comment by thumphries** [comment]
Thanks for your hard work!

I'm noticing with this change that `bazel build ...` fails - part of this is the mangling change (I'm working on getting `master` back to a good state and have turned CI on for pull requests so this stops happening), but after fixing that I still see some failures related to doctest:

```
ERROR: /home/tim/.cache/bazel/_bazel_tim/3dfb6465070eadfbfb3bf40803766bf8/external/haskell_doctest_1828875722/BUILD:17:1: in haskell_library rule @haskell_doctest_1828875722//:Paths_doctest: 
Traceback (most recent call last):
	File "/home/tim/.cache/bazel/_bazel_tim/3dfb6465070eadfbfb3bf40803766bf8/external/haskell_doctest_1828875722/BUILD", line 17
		haskell_library(name = 'Paths_doctest')
	File "/home/tim/.cache/bazel/_bazel_tim/3dfb6465070eadfbfb3bf40803766bf8/external/io_tweag_rules_haskell/haskell/private/haskell_impl.bzl", line 177, in haskell_library_impl
		cc_interop_info(ctx)
	File "/home/tim/.cache/bazel/_bazel_tim/3dfb6465070eadfbfb3bf40803766bf8/external/io_tweag_rules_haskell/haskell/cc.bzl", line 129, in cc_interop_info
		cc_toolchain.ar_executable()
struct field 'ar_executable' is not a function
```

Does it build on your machine? Thoughts? We are using a fairly old Bazel in CI, we should probably catch up there too.

**Comment by mjrussell** [comment]
Dang you beat me to it @thumphries - b28dab8

It is now building on my machine

**Pull request merged**